kotlin-result-coroutines/src/commonMain/kotlin/com/github/michaelbull/result/coroutines/binding/SuspendableBinding.kt

```diff
@@ -3,60 +3,78 @@ package com.github.michaelbull.result.coroutines.binding
+import kotlinx.coroutines.suspendCancellableCoroutine
+import kotlinx.coroutines.launch
+import kotlinx.coroutines.cancel
 import kotlin.contracts.InvocationKind
 import kotlin.contracts.contract
+import kotlin.coroutines.CoroutineContext
+import kotlin.coroutines.resume
 
 /**
  * Suspending variant of [binding][com.github.michaelbull.result.binding].
- * Wraps the suspendable block in a new coroutine scope.
- * This scope is cancelled once a failing bind is encountered, allowing deferred child jobs to be eagerly cancelled.
+ * Wraps the suspendable block in an async job, inheriting the parent coroutine context.
+ * This async job is cancelled once a failing bind is encountered, eagerly cancelling all children.
  */
 public suspend inline fun <V, E> binding(crossinline block: suspend SuspendableResultBinding<E>.() -> V): Result<V, E> {
     contract {
         callsInPlace(block, InvocationKind.EXACTLY_ONCE)
     }
-    val receiver = SuspendableResultBindingImpl<E>()
-
-    return try {
-        coroutineScope {
-            receiver.coroutineScope = this@coroutineScope
+    lateinit var receiver: SuspendableResultBindingImpl<E>
+    return coroutineScope {
+        val parentJob = async {
+            receiver = SuspendableResultBindingImpl(this.coroutineContext)
             with(receiver) { Ok(block()) }
         }
-    } catch (ex: BindCancellationException) {
-        receiver.internalError
+        try {
+            parentJob.await()
+        } catch (ex: BindCancellationException) {
+            receiver.internalError
+        }
     }
 }
 
 internal object BindCancellationException : CancellationException(null)
 
-public interface SuspendableResultBinding<E> {
+public interface SuspendableResultBinding<E> : CoroutineScope {
     public suspend fun <V> Result<V, E>.bind(): V
 }
 
 @PublishedApi
-internal class SuspendableResultBindingImpl<E> : SuspendableResultBinding<E> {
+internal class SuspendableResultBindingImpl<E>(
+    override val coroutineContext: CoroutineContext
+) : SuspendableResultBinding<E> {
 
     private val mutex = Mutex()
     lateinit var internalError: Err<E>
-    var coroutineScope: CoroutineScope? = null
 
+    /**
+     * Suspends the current coroutine while it attempts to unwrap the result value.
+     * If the result is an error, all [binding] block child jobs are cancelled
+     * and [binding] returns result with error of first failed bind.
+     */
     override suspend fun <V> Result<V, E>.bind(): V {
-        return when (this) {
-            is Ok -> value
-            is Err -> {
-                mutex.withLock {
-                    if (::internalError.isInitialized.not()) {
-                        internalError = this
-                    }
+        return suspendCancellableCoroutine {
+            when (this) {
+                is Ok -> it.resume(value)
+                is Err -> setError(this)
+            }
+        }
+    }
+
+    private fun setError(error: Err<E>) {
+        this.launch {
```

Comment:
Why do we need to launch _another_ coroutine just to set our own internal reference to an error?

Reply:
so the reason I had originally used mutex was to avoid racing threads both trying to set the end `Result` failure that the single `SuspendableResultBindingImpl` is holding a reference to. Ideally binding would always return the first failing bind.

`mutex.withLock` ensured that, but it must run inside of a coroutine body, which `suspendCancellableCoroutine` does not provide as we are in a non suspending continuation lambda. Hence the need to have SuspendableBinding launch a new coroutine for each failing bind that is racing so that the lock could be called.

Having read https://kotlinlang.org/docs/shared-mutable-state-and-concurrency.html however, I thought I could actually use `AtomicReference` here instead.

Sadly that only works for kotlin Native 😢 . For multiplatform/common code, it looks like the only solution right now is to add this dependency: https://github.com/Kotlin/kotlinx.atomicfu
Its part of Kotlin org and looks like Roman is a main contributor so I assume its good. So if your cool with that I would add that dependency and get rid of this extra overhead (though in a way trading it for another hehe).

Or we could just not guarantee the failing bind order for binding with coroutines (not trying to sound sarcastic, genuinely think thats an option here as its already quite an edge case).